### ISSUE TITLE
How to use with Declarative Pipeline and docker without dind

### ISSUE THREAD

**Issue by sheerun** [jenkinsci/azure-container-agents-plugin#20]
Hello,

I'm trying to make my declarative pipeline run on azure container instances without employing something like docker in docker. Is there a way to enable something so Jenkinsfile like this could work out of the box on Azure Container Instances?

```
pipeline {
    agent {
        docker { image 'node:7-alpine' }
    }
    stages {
        stage('Test') {
            steps {
                sh 'node --version'
            }
        }
    }
}
```

Here's documentation: https://jenkins.io/doc/book/pipeline/docker/

**Comment by zackliu** [comment]
Docker in docker isn't supported in ACI but supported in ACS

**Comment by zackliu** [comment]
Hi, what your purpose to use Docker inside a container. If you need some specific environment, you can just prepare such a Docker image with JNLP or SSH and use it as agent directly.

**Comment by sheerun** [comment]
I don't need docker inside container (well, I'd like to have it for building images and testing images along services, but that's other matter). The only reason I mentioned dind is that I thought it could indirectly solve my issue by running docker commands inside `steps`, instead of configuring proper image as `agent`. But it turns out it's not possible, so it can be implemented as you say.

This issue is only about ability to run declarative pipeline that I've mentioned, on ACI runtime. I've modified the title of issue to better reflect it.

**Comment by zackliu** [comment]
Sorry, I need to make the issue more clearly. What you want to achieve is if you use the Jenkinsfile above, actually Jenkins will provision an azure container instance with nodejs image to run the following commands. Is my understanding correct?

**Comment by sheerun** [comment]
That's correct :) The point is to be able to run above pipeline unmodified on both ACI and ACS.

**Comment by mitchellmaler** [comment]
I am also wondering the same thing. 

Using Jenkins declarative pipeline can you use the specified ACI cloud but in the pipeline/jenkinsfile specify the image to deploy and use for the job. 

This way we don't have to manually specify the templates but it can be done in the jobs jenkinsfile.

**Comment by nerumo** [comment]
@zackliu you said it works in acs. So if a build requests a part to run in a specific docker image (without jnlp) it should provision a container and run stuff in there?
